language/move-prover/tests/sources/functional/is_signer.move

```diff
@@ -0,0 +1,137 @@
+module 0x42::SimpleIsSigner {
+    use Std::Signer;
+
+    // ---------------
+    // Simple examples
+    // ---------------
+
+    public fun f1_incorrect() {
+        spec { assert Signer::is_signer(@0x7); } // This is unprovable because it is not true in general.
+    }
+
+    public fun f2_incorrect(_account: &signer) {
+        spec { assert Signer::is_signer(@0x7); } // This is unprovable because it is not true in general.
+    }
+
+    public fun f3(account: &signer) {
+        assert(Signer::address_of(account) == @0x7, 1);
+        spec { assert Signer::is_signer(@0x7); } // It's true in general.
+    }
+
+    public fun f4_incorrect(account: &signer) {
+        assert(Signer::address_of(account) == @0x3, 1);
+        spec { assert Signer::is_signer(@0x7); } // This is unprovable because it is not true in general.
+    }
+
+    fun f5() {
+        spec { assert Signer::is_signer(@0x7); } // This is provable because of the "requires" condition.
+    }
+    spec f5 {
+        requires Signer::is_signer(@0x7); // f5 requires this to be true at its callers' sites
+    }
+
+    public fun f6(account: &signer) { // This function produces no error because it satisfies f5's requires condition.
+        assert(Signer::address_of(account) == @0x7, 1);
+        f5();
+    }
+
+    public fun f7_incorrect() {
+        // This function does not satisfy f5's requires condition.
+        f5();
+    }
+
+
+    // ------------------------------
+    // Simple access control examples
+    // ------------------------------
+
+    spec fun hasPermissionAddr(addr: address): bool;
+    fun hasPermission(_account: &signer): bool { true }
+    spec hasPermission {
+        pragma opaque;
+        pragma verify=false;
+        ensures result == hasPermissionAddr(Signer::spec_address_of(_account));
+    }
+
+    public fun g_incorrect(_a: &signer, _b: &signer) {
+        spec {
+            assert exists addr:address: hasPermissionAddr(addr);
+        };
+        // privileged operation
+    }
+
+    public fun g1(a: &signer, _b: &signer) {
+        if(hasPermission(a))
+        {
+            spec {
+                assert exists addr:address: (Signer::is_signer(addr) && hasPermissionAddr(addr));
+            };
+            // privileged operation
+        }
+    }
+
+    public fun g2(a: &signer, _b: &signer) {
+        assert (hasPermission(a), 1);
+        spec {
+            assert exists addr:address: (Signer::is_signer(addr) && hasPermissionAddr(addr));
+        };
+        // privileged operation
+    }
+
+    public fun g3(a: &signer, _b: &signer) {
+        assert (hasPermission(a), 1);
+        helper()
+    }
+
+    public fun helper() {
+        spec {
+            assert exists addr:address: (Signer::is_signer(addr) && hasPermissionAddr(addr));
+        };
+        // privileged operation
+    }
+    spec helper {
+        requires exists addr:address: (Signer::is_signer(addr) && hasPermissionAddr(addr));
+    }
+
```

Comment:
I'm wondering if it makes sense to include a multi-agent test case where the signers have different privileges, and inside the public fun, multiple private functions with different privileged operations are called. Something like that.

Reply:
I think that's a valid scenario. Suppose there is a privileged operation that requires permission P from an agent, and permission Q from another agent. Then, the access control spec could be asserted at around the privileged operation:
```
spec { assert (exists a: addr: (is_signer(a) && hasPermission_P(a)) && (exists a: addr: (is_signer(a) && hasPermission_Q(a)); }
```

Although there are multiple privileged operations requiring multiple permissions, if a single privileged operations does requires multiple permissions, it can be handled in the same way as the case of `g3` and `helper`